tracer/src/Datadog.Trace/Agent/AgentWriter.cs

```diff
@@ -396,21 +397,39 @@ SpanBuffer SwapBuffers()
             }
 
             RunSpanSampler(spans);
-
+            int? spanSamplingPriority = null;
             if (CanComputeStats)
             {
                 spans = _statsAggregator?.ProcessTrace(spans) ?? spans;
                 bool shouldSendTrace = _statsAggregator?.ShouldKeepTrace(spans) ?? true;
                 _statsAggregator?.AddRange(spans);
+                var singleSpanSamplingSpans = new List<Span>(); // TODO maybe we can store this from above?
+
+                for (var i = spans.Offset; i < spans.Count; i++)
+                {
+                    if (spans.Array![i].GetMetric(Metrics.SingleSpanSampling.SamplingMechanism) is not null)
+                    {
+                        singleSpanSamplingSpans.Add(spans.Array![i]);
+                    }
+                }
 
                 // If stats computation determined that we can drop the P0 Trace,
                 // skip all other processing
-                if (!shouldSendTrace)
+                if (!shouldSendTrace && singleSpanSamplingSpans.Count == 0)
                 {
                     Interlocked.Increment(ref _droppedP0Traces);
                     Interlocked.Add(ref _droppedP0Spans, spans.Count);
                     return;
                 }
+                else if (!shouldSendTrace && singleSpanSamplingSpans.Count > 0)
+                {
+                    // we need to set the sampling priority of the chunk to be user keep so the agent handles it correctly
+                    // this will override the TraceContext sampling priority when we do a SpanBuffer.TryWrite
+                    spanSamplingPriority = SamplingPriorityValues.UserKeep;
+                    // note that we aren't incrementing _droppedP0Traces as we aren't dropping the trace entirely
+                    Interlocked.Add(ref _droppedP0Spans, spans.Count - singleSpanSamplingSpans.Count);
```

Comment:
@bouwkast  Actually, let's increment the dropped trace count by 1 here, since we've sampled out the entire trace and then we selectively kept some additional spans

Reply:
Done in 42f725b05edd264298f0395a7cb110cb0a0f9fb9